Rated task rows

PatricKangasmaki opened the pull request:
Separated "Waiting for ratings"- and "Rated tasks" -tables into own components and styled the rated tasks according to figma.

IidaKainu commented on frontend/src/components/TaskTableRated.tsx:
```diff
@@ -0,0 +1,226 @@
+  const getOwnerTask: () => Task[] = () => {
+    const developers = allUsers?.filter(
+      (user) => user.type === RoleType.Developer,
+    );
```
These functions, that are in both TaskTableRated and TaskTableUnrated -files (getOwnerTask, getUnratedTasks?), could be moved to utils-folder

PatricKangasmaki replied:
True 👍 I realized now that I've used functions like these, such as finding missing devs etc.,  in other files as well that could (and should) be moved to the utils-folder. I think I should make a separate pull-request for that since it will be a bit bigger fix. Did other changes in this pr work well?

IidaKainu replied:
A separate pull-request for sounds good if a lot needs to be refactored 👍 However, if changes related to this pr are small, they could be added here (others in the separate pr). Up to you!

juhoeerola commented on frontend/src/components/TaskTableUnrated.module.scss:
```diff
@@ -0,0 +1,11 @@
+@import '../shared.scss';
+
+.tableContainer {
+  margin-bottom: 50px;
+}
```
This might be better defined by the user of these task tables

Also, a trailing newline would be nice for consistency, it seems that your setup leaves that out from scss files.

juhoeerola commented on frontend/src/pages/TaskListPage.tsx:
```diff
@@ -1,9 +1,144 @@
+            <input
+              className={classes(css.search)}
+              placeholder={t('Search for tasks')}
+              onChange={(e: any) => onSearchChange(e.currentTarget.value)}
```
```suggestion
              onChange={(e) => onSearchChange(e.currentTarget.value)}
```

Type inference should work fine

juhoeerola commented on frontend/src/components/TableRatedTaskRow.tsx:
```diff
@@ -0,0 +1,70 @@
+        {task.name}
+      </td>
+      <td className={`styledTd ${classes(css.ratedTd)}`}>
+        {calcAverageTaskValue(task) || 0}
```
Calculating all the necessary averages and sums could be done in one pass over the task ratings

The utils has the basics in https://github.com/Vincit/VISDOM-Roadmapper/blob/20401ca409e6727156efe86307888cad7180f2a3/frontend/src/utils/TaskUtils.ts#L40-L43

Then the sums and averages is simple to get for any dimension

```js
const { sum, count } = ratings.get(dimension) ?? { sum: 0, count: 0 };
const avg = sum / count;
```

juhoeerola commented on frontend/src/components/TaskTableUnrated.tsx:
```diff
@@ -0,0 +1,230 @@
+
+export const TaskTableUnrated: React.FC<{
+  tasks: Task[];
+  nofilter?: boolean;
+  nosearch?: boolean;
```
These are unused

juhoeerola reviewed (changes requested):
I would even go as far as combining the task tables to one, e.g.

```typescript
const TaskTable = /* common implementation */
export const TaskTableUnrated = /* delegates to TaskTable, with necessary adjustments */
export const TaskTableRated   = /* ^ */
```

The main differences are:
- some text content
- the task filtering predicate is flipped
- different header, and task row

PatricKangasmaki commented:
Now it's actually ready for review, I rebased it on master and fixed the merge conflicts after the initial _`PatricKangasmaki marked this pull request as ready for review`_

> These functions, that are in both TaskTableRated and TaskTableUnrated -files (getOwnerTask, getUnratedTasks?), could be moved to utils-folder

Btw, I took your advice and moved the functions into utils-folder in this pr! 😄

IidaKainu commented on frontend/src/components/TaskTable.tsx:
```diff
@@ -255,28 +93,6 @@ export const TaskTable: React.FC<{
-    },
-  ];
-
   const renderUnratedTasks = () => {
```
This function is used for both unrated and rated tasks, so maybe renderTasks would be a more suiting name

IidaKainu reviewed (changes requested):
I think you implemented TaskTable in a neat flexible way! 
Here's just some minor things to consider

juhoeerola reviewed (changes requested):
The task tables ended up nice 👍

juhoeerola commented on frontend/src/components/TaskTable.tsx:
```diff
@@ -301,85 +117,165 @@ export const TaskTable: React.FC<{
+            {label} ({getRenderTaskList().length})
+          </h2>
+          <div>
+            <div>{getRenderTaskList().length && renderUnratedTasks()}</div>
```
Also, is the double `div` needed?

PatricKangasmaki commented on frontend/src/utils/TaskUtils.ts:
```diff
@@ -46,6 +56,16 @@ const averageRatingsByDimension = (
   );
 };
 
+const totalRatingsByDimension = (
+  task: Task,
+): Map<TaskRatingDimension, number> => {
+  const ratings = task.ratings.reduce((result, { value, dimension }) => {
+    const { sum } = result.get(dimension) || { sum: 0 };
+    return result.set(dimension, { sum: sum + value });
+  }, new Map());
+  return new Map(Array.from(ratings).map(([key, { sum }]) => [key, sum]));
+};
```
I'm currently working on this but I'm having troubles with sorting. Currently the function looks like this and it works as planned.

```
export const totalRatingsByDimension = (
  task: Task,
): Map<TaskRatingDimension, {sum: number, count: number}> => {
  const ratings = task.ratings.reduce((result, { value, dimension }) => {
    const { sum, count } = result.get(dimension) || { sum: 0, count: 0 };
    return result.set(dimension, { sum: sum + value, count: count + 1 });
  }, new Map());
  return new Map(
    Array.from(ratings).map(([key, { sum, count }]) => [key, { sum, count }]),
  );
};
```

I'm just wondering how the sorting would be implemented, it used to look this when each calculation had their own functions 
```
case SortingTypes.SORT_AVG_VALUE:
   return sortKeyNumeric(calcAverageTaskValue);
case SortingTypes.SORT_AVG_WORK:
   return sortKeyNumeric(calcAverageTaskWork);
case SortingTypes.SORT_TOTAL_VALUE:
   return sortKeyNumeric(calcTaskValueSum);
case SortingTypes.SORT_TOTAL_WORK:
   return sortKeyNumeric(calcTaskWorkSum);
```

juhoeerola replied:
The individual functions can be kept if needed.

If not used anywhere else, then maybe those can be moved to the sorting module as local non-exported functions.
Or the ones that are simple enough, could be inlined.

PatricKangasmaki commented:
All the suggestions are now added in the code 👍 I decided to leave the functions used to sort value/work columns in the taskUtils as separate non exported functions and use the new totalRatingsByDimension in calculations

juhoeerola commented on frontend/src/utils/TaskUtils.ts:
```diff
@@ -269,3 +334,30 @@ export const taskAwaitsRatings = (task: Task, userInfo?: UserInfo) => {
+  tasks: Task[],
+  allUsers: RoadmapUser[],
+  allCustomers: Customer[],
+) => Task[] | undefined = (tasks, allUsers, allCustomers) => {
+  const developers = allUsers.filter(
+    (user) => user.type === RoleType.Developer,
+  );
+  const unrated = tasks.filter((task) => {
+    const ratingIds = task.ratings.map((rating) => rating.createdByUser);
+
+    const customersMissing = allCustomers?.some((customer) => {
```
```suggestion
) => Task[] = (tasks, allUsers, allCustomers) => {
  const developers = allUsers.filter(
    (user) => user.type === RoleType.Developer,
  );
  const unrated = tasks.filter((task) => {
    const ratingIds = task.ratings.map((rating) => rating.createdByUser);

    const customersMissing = allCustomers.some((customer) => {
```

juhoeerola commented on frontend/src/components/TaskTable.tsx:
```diff
@@ -301,85 +117,155 @@ export const TaskTable: React.FC<{
+    if (type === RoleType.Admin) {
+      if (userInfo && currentRoadmap && allUsers)
+        setTaskList(
+          unratedProductOwnerTasks(tasks, allUsers, currentRoadmap?.customers),
```
```suggestion
          unratedProductOwnerTasks(tasks, allUsers, currentRoadmap.customers),
```

juhoeerola commented on frontend/src/components/TaskTable.tsx:
```diff
@@ -301,85 +117,155 @@ export const TaskTable: React.FC<{
+    const type = getType(userInfo?.roles, currentRoadmap?.id);
+    if (type === RoleType.Admin) {
+      if (userInfo && currentRoadmap && allUsers) {
+        const unratedTasks = unratedProductOwnerTasks(tasks, allUsers, currentRoadmap?.customers)
```
```suggestion
        const unratedTasks = unratedProductOwnerTasks(tasks, allUsers, currentRoadmap.customers)
```

juhoeerola commented on frontend/src/components/TableRatedTaskRow.tsx:
```diff
@@ -0,0 +1,84 @@
+    });
+  };
+
+  return (
+    <tr
+      className={classes(css.tableRow, {
+        [css.hoverRow]: hovered,
+        [css.completedRow]: task.completed,
+      })}
+    >
+      <td className={classes(css.ratedTitle)}>
+        {task.completed && <DoneAllIcon className={classes(css.doneIcon)} />}
+        {task.name}
+      </td>
+      <td className={classes(css.ratedTd)}>
+        {numFormat(calcTaskValues(task).sum / calcTaskValues(task).count)}
+      </td>
```
I'm a bit uneasy with the amount of duplicate calls here

What I was suggesting earlier by doing the calculation once (sorry if I was unclear):

> Calculating all the necessary averages and sums could be done in one pass over the task ratings

> could also be used in the `TableRatedTaskRow` directly, instead of all the `calc*` calls

Something like this could work:

```suggestion
  const totalRatings = totalRatingsByDimension(task);
  const noRatings = { sum: 0, count: 1 };
  const valueRatings =
    totalRatings.get(TaskRatingDimension.BusinessValue) ?? noRatings;
  const workRatings =
    totalRatings.get(TaskRatingDimension.RequiredWork) ?? noRatings;

  return (
    <tr
      className={classes(css.tableRow, {
        [css.hoverRow]: hovered,
        [css.completedRow]: task.completed,
      })}
    >
      <td className={classes(css.ratedTitle)}>
        {task.completed && <DoneAllIcon className={classes(css.doneIcon)} />}
        {task.name}
      </td>
      <td className={classes(css.ratedTd)}>
        {numFormat(valueRatings.sum / valueRatings.count)}
```

juhoeerola reviewed (changes requested):
Found few possible improvements

Otherwise looks great 👍

juhoeerola commented on frontend/src/components/PlannerChart.tsx:
```diff
@@ -82,7 +82,11 @@ export const PlannerChart: React.FC<{
+            TaskRatingDimension.BusinessValue,
+          ) ?? { sum: 0, count: 1 };
+          const value = valueRatings.sum ?? 0;
+          // totalRatingsByDimension(task).get(TaskRatingDimension.BusinessValue).sum ?? 0
```
Is this still in progress, or just left over?

PatricKangasmaki replied:
Forgot to delete that comment, everything is ready

juhoeerola reviewed (approved):
👍 

If the PlannerChart is ready, this is good, just drop the commented out code before merging
